anchor/src/processor.rs

```diff
@@ -33,12 +71,57 @@ fn process_initialize(program_id: &Pubkey, accounts_raw: &[AccountInfo]) -> Prog
 fn process_deposit(
     program_id: &Pubkey,
     accounts_raw: &[AccountInfo],
-    _amount: StLamports,
+    amount: StLamports,
 ) -> ProgramResult {
     let accounts = DepositAccountsInfo::try_from_slice(accounts_raw)?;
-    let _anchor = deserialize_anchor(program_id, accounts.anchor)?;
+    if amount == StLamports(0) {
+        msg!("Amount must be greater than zero");
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let anchor = deserialize_anchor(
+        program_id,
+        accounts.anchor,
+        accounts.lido.key,
+        accounts.to_reserve_account.key,
+    )?;
+
+    // Check if the mint account is the same as the one stored in Anchor.
+    anchor.check_mint(accounts.b_sol_mint.key)?;
+
+    // Transfer `amount` StLamports to the reserve.
+    invoke(
+        &spl_token::instruction::transfer(
+            &spl_token::id(),
+            accounts.from_account.key,
+            accounts.to_reserve_account.key,
+            accounts.user_authority.key,
+            &[],
+            amount.0,
+        )?,
+        &[
+            accounts.from_account.clone(),
+            accounts.to_reserve_account.clone(),
+            accounts.user_authority.clone(),
+            accounts.spl_token.clone(),
+        ],
+    )?;
+
+    let lido = Lido::deserialize_lido(accounts.lido_program.key, accounts.lido)?;
+
+    // Use Lido's exchange rate (`st_sol_supply / sol_balance`) to compute the
+    // amount of BLamports to mint.
+    let amount = BLamports(lido.exchange_rate.exchange_sol(Lamports(amount.0))?.0);
```

Comment:
This is the wrong way around, when you deposit stSOL, we want to know how much SOL that is worth, and mint the same amount of bSOL. So it should be

```rust
let sol_value = lido.exchange_rate.exchange_st_sol(amount)?;
let b_sol_amount = BLamports(sol_value.0);
```

The types become simpler too!

Reply:
Oh, that's correct. Changed it